### PR TITLE
gh-112213: Add @critical_section target directive to Argument Clinic

### PR DESCRIPTION
**Pull request by corona10**
* Issue: gh-112213

### CODE REVIEW

**Comment by corona10** [comment]
@colesbury Do we have to support criticalsection2 API too for two object cases?

File: Tools/clinic/clinic.py [hunk 1]
```diff
@@ -1866,8 +1866,15 @@ def render_function(
...
-            data.unlock.append('Py_END_CRITICAL_SECTION();')
+            if len(f.target_critical_section) == 0:
+                data.lock.append('Py_BEGIN_CRITICAL_SECTION({self_name});')
+                data.unlock.append('Py_END_CRITICAL_SECTION();')
+            elif len(f.target_critical_section) == 1:
+                data.lock.append('Py_BEGIN_CRITICAL_SECTION({target_critical_section});')
+                data.unlock.append('Py_END_CRITICAL_SECTION();')
+            else:
+                data.lock.append('Py_BEGIN_CRITICAL_SECTION2({target_critical_section});')
+                data.unlock.append('Py_END_CRITICAL_SECTION2();')
```
Comment (AlexWaygood) [inline Tools/clinic/clinic.py]:
```suggestion
            match len(f.target_critical_section):
                case 0:
                    lock = 'Py_BEGIN_CRITICAL_SECTION({self_name});'
                    unlock = 'Py_END_CRITICAL_SECTION();'
                case 1:
                    lock = 'Py_BEGIN_CRITICAL_SECTION({target_critical_section});'
                    unlock = 'Py_END_CRITICAL_SECTION();'
                case _:
                    lock = 'Py_BEGIN_CRITICAL_SECTION2({target_critical_section});'
                    unlock = 'Py_END_CRITICAL_SECTION2();'
            data.lock.append(lock)
            data.unlock.append(unlock)
```

File: Tools/clinic/clinic.py [hunk 2]
```diff
@@ -5288,7 +5298,11 @@ def at_classmethod(self) -> None:
             fail("Can't set @classmethod, function is not a normal callable")
         self.kind = CLASS_METHOD
 
-    def at_critical_section(self) -> None:
+    def at_critical_section(self, *args: str) -> None:
+        if len(args) > 2:
+            fail("Up to 2 critical section variables are supported")
+        for t in args:
+            self.target_critical_section.append(t)
```
Comment (AlexWaygood) [inline Tools/clinic/clinic.py]:
```suggestion
        self.target_critical_section.extend(args)
```

File: Tools/clinic/clinic.py [hunk 3]
```diff
@@ -5514,7 +5528,7 @@ def state_modulename_name(self, line: str) -> None:
 
         self.function = Function(name=function_name, full_name=full_name, module=module, cls=cls, c_basename=c_basename,
                                  return_converter=return_converter, kind=self.kind, coexist=self.coexist,
-                                 critical_section=self.critical_section)
+                                 critical_section=self.critical_section, target_critical_section=self.target_critical_section)
```
Comment (AlexWaygood) [inline Tools/clinic/clinic.py]: Let's wrap new code to ~90 characters per line where possible :)

```suggestion
                                 critical_section=self.critical_section,
                                 target_critical_section=self.target_critical_section)
```

**Review by AlexWaygood — COMMENTED** [review]
Some small suggestions:

File: Tools/clinic/clinic.py (hunk 1 above)
Comment (corona10) [reply Tools/clinic/clinic.py]: Thanks! I sometime forget the existence of pattern matching!
This is a good use case.

**Review by AlexWaygood — APPROVED** [review]
The Python code LGTM! I'll leave it to the other two to review the generated C code :D

**Review by colesbury — APPROVED** [review]
Great!

**Pull request merged**